### PR TITLE
Add freezing pattern for fusing int4 mm kernel

### PR DESCRIPTION
**Pull request by mergennachin**
Summary:

Followup to
https://github.com/pytorch/executorch/pull/15813#issuecomment-3530199109

This PR adds INT4 GPU matmul fusion patterns to Inductor's freezing passes, enabling automatic fusion of multiple _weight_int4pack_mm operations that share the same input (e.g., Q/K/V projections in attention layers) into a single fused matmul followed by a split. The pattern concatenates packed INT4 weights on dim=0 and scale_zeros on dim=1, reducing 2-3 separate INT4 matmul calls to just one, which improves performance for INT4 quantized models. The implementation supports CUDA and MPS backends, validates weight compatibility (same K dimension, matching dtypes), and registers patterns for common group sizes (32, 64, 128, 256). Includes a dedicated test file with 6 tests covering 2-way and 3-way fusion, various group sizes, and negative cases where fusion should not occur.

Run locally on A100
4%-5% latency speed, within 95% confidence interval: [1.0415x, 1.0514x] compared to no-freezing mode.

Memory increased quite a bit, but that's related to freezing in general (constant folding) and existed before this optimizaiton.

Appendix: 
Here's the model loading script: https://gist.github.com/mergennachin/d6a0021d0240c60e96af0370802ed594
Benchmark code: https://gist.github.com/mergennachin/360886d0815bd3d9d58b3d0106877756
Benchmark util code: https://gist.github.com/mergennachin/81503405ffa8a90c4c1e245aefd6afaf
Benchmark code with Statistical significance testing: https://gist.github.com/mergennachin/d2430dda0469262ed90d1ae0a3f7d995

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_inductor_freezing_int4mm.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+                return mod(inp)
+
+            with torch.no_grad():
+                out, code = run_and_get_code(foo, model, x)
+
+        # Verify output correctness
+        for i in range(3):
+            self.assertEqual(ref[i], out[i])
+
+        # Count INT4 matmul calls - should be 1 (fused) not 3
```
Comment (desertfire) [inline test/inductor/test_inductor_freezing_int4mm.py]: There is an existing utility function, `FileCheck().check_cout`, for this kind of checking, see https://github.com/pytorch/pytorch/blob/c3c11b794173a4e6641f27ead4ce20ea203a7ac5/test/inductor/test_torchinductor.py#L11964

File: test/inductor/test_inductor_freezing_int4mm.py [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+        with config.patch(
+            {
+                "freezing": True,
+                "freezing_discard_parameters": False,
+                "implicit_fallbacks": True,
+            }
+        ):
+
+            @torch.compile()
+            def foo(mod, inp):
```
Comment (desertfire) [inline test/inductor/test_inductor_freezing_int4mm.py]: Nit: you can just do `foo = torch.compile(mod)`

File: test/inductor/test_inductor_freezing_int4mm.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+        # Get reference output
+        with torch.no_grad():
+            ref = model(x)
+
+        # Compile with freezing
+        with config.patch(
+            {
+                "freezing": True,
+                "freezing_discard_parameters": False,
+                "implicit_fallbacks": True,
```
Comment (desertfire) [inline test/inductor/test_inductor_freezing_int4mm.py]: Curious, is this really needed?

**Review by desertfire — APPROVED** [review]
LGTM. Please fix lint errors.

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 4]
```diff
@@ -121,6 +121,177 @@ def register_binary_folding_pattern(pattern, extra_check=_return_true):
...
+            ):
+                return False
+            # Dtype must match for concatenation
+            if sz_val.dtype != first_sz.dtype:
+                return False
+
+        return True
+
+    # Register patterns for common group_size values
+    for gs in [32, 64, 128, 256]:
```
Comment (eellison) [inline torch/_inductor/fx_passes/freezing_patterns.py]: Can you check that the sizes are necessary ? We erase ints in the pattern for searching, then we retracing with actual sizes.

**Review by eellison — COMMENTED** [review]
looks good but one comment
